File: src/decoders/menu.ts
import { type Menu, MealKind } from "~/models";
import { decodePronoteDate } from "./pronote-date";
import { decodeMeal } from "./meal";

export const decodeMenu = (menu: any): Menu => {
  let entries = menu.ListeRepas.V.reduce((acc: any, meal: any) => {
    let key: string;

    switch (meal.G) {
      case MealKind.Lunch:
        key = "lunch";
        break;
      case MealKind.Dinner:
        key = "dinner";
        break;
      default:
        throw new Error("unreachable");
    }

    acc[key] = decodeMeal(meal);
    return acc;
  }, {});

  return {
    date: decodePronoteDate(menu.Date.V),
    ...entries
  };
};

File: library/src/commonMain/kotlin/ink/literate/pawnote/decoders/menu.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.models.Meal
import ink.literate.pawnote.models.MealKind
import ink.literate.pawnote.models.Menu

import kotlinx.serialization.json.*

fun decodeMenu (menu: JsonObject): Menu {
    var lunch: Meal? = null
    var dinner: Meal? = null

    menu["ListeRepas"]!!.jsonObject["V"]!!.jsonArray.map { meal ->
        val data = decodeMeal(meal.jsonObject)

        when (MealKind.fromInt(meal.jsonObject["G"]!!.jsonPrimitive.int)) {
            MealKind.Lunch -> lunch = data
            MealKind.Dinner -> dinner = data
        }
    }

    return Menu(
        date = decodePronoteDate(menu["Date"]!!.jsonObject["V"]!!.jsonPrimitive.content),
        lunch = lunch,
        dinner = dinner
    )
}